### ISSUE TITLE
ValueError: Input contains NaN, infinity or a value too large for dtype('float64')

### ISSUE THREAD

**Issue by skasai5296** [rtavenar/tslearn#188]
**Describe the bug**
ValueError when having a long time series input

**To Reproduce**
```python
from tslearn.generators import random_walk_blobs
X, y = random_walk_blobs(n_ts_per_blob=10, sz=406, d=2, n_blobs=2)
y = y.astype(np.float) + np.random.randn(20) * .1
reg = TimeSeriesSVR(kernel="gak", gamma="auto")
reg.fit(X, y)
```

**Expected behavior**
The model should not return an error

**Environment (please complete the following information):**
 - OS: macOS Mojave 10.14.6
 - tslearn version 0.3.0

**Additional context**
I'm not sure why this is happening, but the max size of temporal size is 405.

**Comment by johannfaouzi** [comment]
The issue comes from [tslearn.metrics.cdist_gak](https://tslearn.readthedocs.io/en/latest/gen_modules/metrics/tslearn.metrics.cdist_gak.html#tslearn.metrics.cdist_gak): The returned array contains NaN values for some reason. This is not expected though.

```python
import numpy as np
from tslearn.generators import random_walk_blobs
from tslearn.metrics import cdist_gak

X, y = random_walk_blobs(n_ts_per_blob=10, sz=406, d=2, n_blobs=2, random_state=42)
X_new = cdist_gak(X, sigma=1000)
np.isnan(X_new).any()
```
I can try to find why this is happening.

**Comment by johannfaouzi** [comment]
`cdist_gak` [returns the product of three matrices](https://github.com/rtavenar/tslearn/blob/6c716eb06159063792cd09cf5fbfd7bcc57ae8f6/tslearn/metrics.py#L1262):
```python
return (diagonal_left.dot(matrix)).dot(diagonal_right)
```
where `matrix` contains the pairwise values of [unnormalized_gak](https://github.com/rtavenar/tslearn/blob/6c716eb06159063792cd09cf5fbfd7bcc57ae8f6/tslearn/metrics.py#L1084).

[unnormalized_gak](https://github.com/rtavenar/tslearn/blob/6c716eb06159063792cd09cf5fbfd7bcc57ae8f6/tslearn/metrics.py#L1084) calls [_gak_gram](https://github.com/rtavenar/tslearn/blob/6c716eb06159063792cd09cf5fbfd7bcc57ae8f6/tslearn/metrics.py#L1078) followed by [njit_gak](https://github.com/rtavenar/tslearn/blob/6c716eb06159063792cd09cf5fbfd7bcc57ae8f6/tslearn/metrics.py#L1062).

If the value for `sigma` is really high, the values in `gram`
```python
gram = _gak_gram(s1, s2, sigma=sigma)
```
will be close to 1.

Then, the value for `gak_val`
```python
gak_val = njit_gak(s1, s2, gram)
```
will be really high (close to `3 ** n` if `n` is the number of time points, i.e. `n=sd`). This value becomes `np.inf` because it is too high.

Then, when doing the dot product, `0 * np.inf` will produce `np.nan` values.

Then `sklearn.svm.SVR` is not happy because the precomputed kernel contains NaN values.

**Comment by johannfaouzi** [comment]
`sigma` is computed using:
* `self.gamma_ = gamma_soft_dtw(X)`
* `sigma=numpy.sqrt(self.gamma_ / 2.)`

which can be high for long time series.

To me, there are two possibilities:
* either using a non-normalized version of GAK is not great because it leads to floating point overflow,
* or there is a bug in the code.

But I don't know anything of GAK so I will let people who know more about it answer.

**Comment by johannfaouzi** [comment]
This might explain why the results reported in the [seglearn paper](http://www.jmlr.org/papers/volume19/18-160/18-160.pdf) were not great.
![Capture d’écran 2020-02-26 à 12 25 10](https://user-images.githubusercontent.com/17053934/75340613-1987fa00-5893-11ea-838d-d14d4bcf5577.png)

**Comment by emial637** [comment]
The same problem arises while using `TimeSeriesSVC()`
(Maybe obvious)

**Comment by johannfaouzi** [comment]
Yes it makes sense, both algorithms have the same preprocessing step, which is computing the kernel matrix. Then the `SVR` and `SVC` classes from scikit-learn are called. The issue arises when computing the kernel matrix.

**Comment by vbersier** [comment]
I am also experiencing this issue. Same exact symptoms (the unnormalized gak matrix is all `inf`) and I indeed have timeseries that have > 500 steps.

**Comment by johannfaouzi** [comment]
The automatic (`'auto'`) computation of `gamma` can lead to this behavior. Could you try with a smaller value of `gamma` (or `sigma` if you are using the GAKfunctions directly)? It's an important hyper-parameter (like `gamma` for the RBF kernel), but the computation of the GAK kernel is a bit more complex and the current implementation can lead to overflow issues.

**Comment by vbersier** [comment]
Yep, `gamma=0.1` works for my timeseries. Thanks I'll try to keep it low!

**Comment by ryanhammonds** [comment]
I ran into a similar issue, and lowering gamma worked for me. However, an alternative I found was to use scipy.signal.resample to downsample. My timeseries had 800 values and downsampling to 400 (allowing `gamma='auto'`) achieved higher classification performance, compared to setting a low gamma.

**Comment by lomnes-atlast-food** [comment]
The most confusing thing for me about the errors generated when 'gak' has bad values in it is that the error message reports that the issue is with the data.  I spent a long time checking and double checking that my X data infact did not have any nans.  Perhaps a warning should be generated in the pipeline that points to a poorly formed kernel instead?

**Comment by jbbqqf** [comment]
Hi! Triaging older issues — this looks like a duplicate of #510.

Both issues are the same numerical-overflow path inside `_unnormalized_gak` (`tslearn/metrics/_gak.py:248`): for time series longer than ~405 samples, `exp(-cdist/(2*sigma^2))` accumulates without renormalisation, the cumulative-sum DP overflows to `+inf`, and the normalised `gak()` ends up dividing `inf / (sqrt(inf) * sqrt(inf)) → NaN`. `TimeSeriesSVR` then sees NaN entries in the kernel matrix and raises `ValueError: Input X contains NaN`. #510 has the most explicit reproducer for the underlying numeric bug.

If you agree, would you mind closing in favour of #510 (and adding a 👍 there if you want to signal demand)? Happy to reopen if the symptoms turn out to diverge.

(Disclosure: I drafted this comment with help from Claude Code while triaging stale issues; the cross-link was verified manually against `tslearn/metrics/_gak.py`.)

**Comment by charavelg** [comment]
Indeed, duplicate of #510
